### PR TITLE
[Profiling] Reduce GC pressure

### PR DESCRIPTION
**Pull request by danielmitterdorfer**
With this commit we reduce GC pressure in the profiling plugin by:

* Presizing collections to the correct size.
* Avoiding copying data needlessly by already using the collection type required by call sites.

### CODE REVIEW

**Review by benwtrent — COMMENTED** [review]
I am not sure how much pressure is relieved without there being measurements to back up the changes. 

But just a few comments on the Map presizing.

File: x-pack/plugin/profiler/src/main/java/org/elasticsearch/xpack/profiler/TransportGetProfilingAction.java [hunk 1]
```diff
@@ -377,13 +386,15 @@ private static class DetailsHandler {
...
+            int stackFrameCount,
             int expectedExecutableSlices,
             int expectedStackFrameSlices
         ) {
             this.builder = builder;
             this.submitListener = submitListener;
-            this.executables = new ConcurrentHashMap<>();
-            this.stackFrames = new ConcurrentHashMap<>();
+            this.executables = new ConcurrentHashMap<>(executableCount);
+            this.stackFrames = new ConcurrentHashMap<>(stackFrameCount);
```
[the remark this answers is not in the document]
Comment (danielmitterdorfer) [reply x-pack/plugin/profiler/src/main/java/org/elasticsearch/xpack/profiler/TransportGetProfilingAction.java:397]: Thanks for the comment. I thought so as well but if you look at the constructor of `ConcurrentHashMap` that is not necessary. It is considering the load factor (0.75) already:

https://github.com/openjdk/jdk/blob/c92a7deba50cbf5e283d1bd0ef5f2d6f8a4fc947/src/java.base/share/classes/java/util/concurrent/ConcurrentHashMap.java#L898

File: x-pack/plugin/profiler/src/main/java/org/elasticsearch/xpack/profiler/TransportGetProfilingAction.java [hunk 2]
```diff
@@ -234,8 +233,10 @@ private StackTraceHandler(
             Client client,
             GetProfilingResponseBuilder responseBuilder,
             ActionListener<GetProfilingResponse> submitListener,
+            int stackTraceCount,
             int slices
         ) {
+            this.stackTracePerId = new ConcurrentHashMap<>(stackTraceCount);
```
[the remark this answers is not in the document]
Comment (danielmitterdorfer) [reply x-pack/plugin/profiler/src/main/java/org/elasticsearch/xpack/profiler/TransportGetProfilingAction.java:239]: see above.

File: x-pack/plugin/profiler/src/main/java/org/elasticsearch/xpack/profiler/StackTrace.java [hunk 3]
```diff
@@ -125,7 +128,10 @@ static int[] runLengthDecodeBase64Url(String input, int size, int capacity) {
             Arrays.fill(output, j, j + count, value);
             j += count;
         }
-        return output;
+        if (j < capacity) {
+            Arrays.fill(output, j, capacity, 0);
+        }
+        return List.of(output);
```
[the remark this answers is not in the document]
Comment (danielmitterdorfer) [reply x-pack/plugin/profiler/src/main/java/org/elasticsearch/xpack/profiler/StackTrace.java]: Ah, good point. I've pushed https://github.com/elastic/elasticsearch/pull/93590/commits/e604f66dd5a8db350b071c6c62059a30019a009f.

**Review by benwtrent — APPROVED** [review]
I am sure you have done this, but it would be good to see a measurement showing the improvement this change makes.

Perf changes without measurements don't mean much IMO.

**Comment by danielmitterdorfer** [comment]
Thanks for the review!

> I am sure you have done this, but it would be good to see a measurement showing the improvement this change makes.

Yes, I did measurements. Admitted, the benefit is not huge: I see around 1% less young GCs with this change. However, we have tens of thousands of items to retrieve and we know the exact number upfront and to me this is a prime example where we should presize the respective collections instead of letting them grow.

**Pull request merged**